src/libsystemd-network/sd-ndisc.c

```diff
@@ -420,6 +421,277 @@ _public_ int sd_ndisc_start(sd_ndisc *nd) {
+
+        log_ndisc("Started IPv6 Router Advertisment daemon");
+
+        return 0;
+}
+
+_public_ int sd_radv_set_ifindex(sd_radv *ra, int ifindex) {
+        assert_return(ra, -EINVAL);
+        assert_return(ifindex >= -1, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->ifindex = ifindex;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_mac(sd_radv *ra, const struct ether_addr *mac_addr) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        if (mac_addr)
+                ra->mac_addr = *mac_addr;
+        else
+                zero(ra->mac_addr);
+
+        return 0;
+}
+
+_public_ int sd_radv_set_mtu(sd_radv *ra, uint32_t mtu) {
+        assert_return(ra, -EINVAL);
+        assert_return(mtu >= 1280, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->mtu = mtu;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_hop_limit(sd_radv *ra, uint8_t hop_limit) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->hop_limit = hop_limit;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_router_lifetime(sd_radv *ra, uint32_t router_lifetime) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->lifetime = router_lifetime;
+
+        if (ra->lifetime == 0)
+                (void) sd_radv_set_preference(ra, SD_NDISC_PREFERENCE_MEDIUM);
+
+        return 0;
+}
+
+_public_ int sd_radv_set_managed_information(sd_radv *ra, bool managed) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        if (managed)
+                ra->flags |= ND_RA_FLAG_MANAGED;
+        else
+                ra->flags &= ~ND_RA_FLAG_MANAGED;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_other_information(sd_radv *ra, bool other) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        if (other)
+                ra->flags |= ND_RA_FLAG_OTHER;
+        else
+                ra->flags &= ~ND_RA_FLAG_OTHER;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_preference(sd_radv *ra, unsigned preference) {
+        int r = 0;
+
+        assert_return(ra, -EINVAL);
+        assert_return(IN_SET(preference,
+                             SD_NDISC_PREFERENCE_LOW,
+                             SD_NDISC_PREFERENCE_MEDIUM,
+                             SD_NDISC_PREFERENCE_HIGH), -EINVAL);
+
+        /* RFC 4191, Section 2.2, "...If the Router Lifetime is zero, the
+           preference value MUST be set to (00) by the sender..." */
+        if (ra->lifetime == 0) {
+                preference = SD_NDISC_PREFERENCE_MEDIUM;
+                r = -EINVAL;
```

Comment:
It seems strange to modify the internal state and yet return an error. I think it'd be cleaner to ensure that preference is set to medium whenever the lifetime is set to 0, and just return an error here.

Also, the error should not be EINVAL, but something different to allow the caller to differentiate this case (ETIME?).